### PR TITLE
[feature](Nereids): Enable the costAndEnforcerJob

### PR DESCRIPTION
**Pull request by jackwener**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Enable the costAndEnforcerJob

Fix some bug of enforcer.

polish property name and method

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/cascades/CostAndEnforcerJob.java [hunk 1]
```diff
@@ -66,18 +67,18 @@ public CostAndEnforcerJob(GroupExpression groupExpression, JobContext context) {
...
-            if (!childGroup.isHasCost()) {
-                // TODO: interim solution
-                pushTask(new CostAndEnforcerJob(this.groupExpression, context));
-                pushTask(new OptimizeGroupJob(childGroup, context));
-                childGroup.setHasCost(true);
-                return;
-            }
-        }
-    }
+    //    @Override
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/cascades/CostAndEnforcerJob.java:70]: Should remove these code.

File: fe/fe-core/src/test/java/org/apache/doris/nereids/jobs/CostAndEnforcerJobTest.java [hunk 2]
```diff
@@ -55,60 +49,69 @@ public class CostAndEnforcerJobTest {
...
+        List<SlotReference> t3Output = scan3.getOutput().stream().map(slot -> (SlotReference) slot)
+                .collect(Collectors.toList());
+        outputs.add(t1Output);
+        outputs.add(t2Output);
+        outputs.add(t3Output);
+    }
+
     @Test
-    public void testExecute(@Mocked LogicalProperties logicalProperties) {
+    public void testExecute() {
```
Comment (jackwener) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/jobs/CostAndEnforcerJobTest.java:78]: Clear this UT

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecHash.java [hunk 3]
```diff
@@ -68,7 +68,7 @@ public boolean satisfy(DistributionSpec other) {
         // TODO: need consider following logic whether is right, and maybe need consider more.
 
         // Current shuffleType is LOCAL/AGG, allow if current is contained by other
-        if (shuffleType == ShuffleType.LOCAL && spec.shuffleType == ShuffleType.AGG) {
+        if (shuffleType == ShuffleType.LOCAL || spec.shuffleType == ShuffleType.AGG) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecHash.java:71]: better to add an example to explain why agg is ok

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java [hunk 4]
```diff
@@ -31,12 +44,35 @@ public class LogicalOlapScanToPhysicalOlapScan extends OneImplementationRuleFact
...
     public Rule build() {
         return logicalOlapScan().then(olapScan ->
-                // TODO: olapScan should get (OlapTable);
-                new PhysicalOlapScan(
-                    (OlapTable) olapScan.getTable(),
-                    olapScan.getQualifier(),
-                    Optional.empty(),
-                    olapScan.getLogicalProperties())
+            new PhysicalOlapScan(
+                (OlapTable) olapScan.getTable(),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java:48]: add a Override function in LogicalOlapScan, and then remove cast on `olapScan.getTable()`
```java
@Override
OlapTable getTable();
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java [hunk 5]
```diff
@@ -31,12 +44,35 @@ public class LogicalOlapScanToPhysicalOlapScan extends OneImplementationRuleFact
...
+                olapScan.getLogicalProperties())
         ).toRule(RuleType.LOGICAL_OLAP_SCAN_TO_PHYSICAL_OLAP_SCAN_RULE);
     }
+
+    private DistributionSpec convertDistribution(LogicalOlapScan olapScan) {
+        DistributionInfo distributionInfo = olapScan.getTable().getDefaultDistributionInfo();
+        if (distributionInfo instanceof HashDistributionInfo) {
+            HashDistributionInfo hashDistributionInfo = (HashDistributionInfo) distributionInfo;
+
+            List<SlotReference> output = olapScan.getOutput().stream().map(slot -> (SlotReference) slot)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java:61]:
```suggestion
            List<SlotReference> output = olapScan.getOutput().stream().map(SlotReference.class:cast)
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java [hunk 6]
```diff
@@ -71,21 +71,21 @@ public Void visit(Plan plan, PlanContext context) {
...
+            List<PhysicalProperties> propertiesForShuffle = Lists.newArrayList(
+                    new PhysicalProperties(new DistributionSpecHash(onClauseUsedSlots.first, ShuffleType.JOIN)),
+                    new PhysicalProperties(new DistributionSpecHash(onClauseUsedSlots.second, ShuffleType.JOIN)));
+
             requestPropertyToChildren.add(propertiesForShuffle);
         }
+        // for broadcast join
         if (!JoinUtils.onlyShuffle(hashJoin)) {
+            List<PhysicalProperties> propertiesForBroadcast = Lists.newArrayList(
+                    new PhysicalProperties(),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java:86]: i think it is better to create any distribution properties explicitly. such as
```
new PhysicalProperties(DistributionSpecAny.getInstance())
```
or
```
PhysicalProperties.CreateDistributionSpecAnyInstance();
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 7]
```diff
@@ -78,7 +78,8 @@ private static boolean isEqualTo(List<SlotReference> leftSlots, List<SlotReferen
             return false;
         }
 
-        return Utils.equalsIgnoreOrder(leftUsed, leftSlots) || Utils.equalsIgnoreOrder(rightUsed, rightSlots);
+        return (new HashSet<>(leftSlots).containsAll(leftUsed) && new HashSet<>(rightSlots).containsAll(rightUsed))
+                || (new HashSet<>(leftSlots).containsAll(rightUsed) && new HashSet<>(rightSlots).containsAll(leftUsed));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:82]:
```suggestion
        Set<> leftSlotsSet = new HashSet<>(leftSlots);
        Set<> rightSlotsSet = new HashSet<>(rightSlots);
        ...
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 8]
```diff
@@ -112,6 +113,7 @@ public static Pair<List<SlotReference>, List<SlotReference>> getOnClauseUsedSlot
             }
         }
 
+        Preconditions.checkArgument(childSlots.first.size() == childSlots.second.size());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:116]:
```suggestion
        Preconditions.checkState(childSlots.first.size() == childSlots.second.size());
```
https://guava.dev/releases/19.0/api/docs/com/google/common/base/Preconditions.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java [hunk 9]
```diff
@@ -17,35 +17,47 @@
...
-        this.distributionSpec = DistributionSpecAny.getInstance();
+        this.distributionSpec = DistributionSpecAny.INSTANCE;
     }
 
     public PhysicalProperties(DistributionSpec distributionSpec, OrderSpec orderSpec) {
         this.distributionSpec = distributionSpec;
         this.orderSpec = orderSpec;
     }
 
+    public static PhysicalProperties createHash(DistributionSpec distributionSpec) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java:57]: this function name is `createHash`, so i think the parameter should be `DistributionSpecHash`

**Review by 924060929 — APPROVED** [review]
LGTM

**Pull request merged**